File: app/channels/document_channel.rb
class DocumentChannel < ApplicationCable::Channel
  def subscribed
    @document = Document.find(params[:document_id])
    @client_id = params[:client_id]

    Rails.logger.info("DocumentChannel subscribed, should receive stream from:")
    Rails.logger.info("  #{broadcast_key}")
    Rails.logger.info("  #{client_key}")

    stream_from broadcast_key
    stream_from client_key
  end

  def request_operations_since(data)
    Rails.logger.info("DocumentChannel operations since #{data}")
    send_operations_since(data["version"])
  end

  # { type: ActionType, payload: ActionPayload, version: number }
  def submit(data)
    Rails.logger.info(
      "DocumentChannel submit #{data} with handle on #{@document}",
    )

    # always apply on fresh @document
    @document.reload

    operation = @document.generate_operation(data)

    Rails.logger.info("DocumentChannel operation #{operation}")

    Operation.transaction do
      operation.save!
      @document.apply(operation)
      @document.save!
    end

    ActionCable.server.broadcast(
      client_key,
      { type: "ack", message: operation },
    )
    ActionCable.server.broadcast(
      broadcast_key,
      { type: "op", client_id: @client_id, message: [operation] },
    )
  rescue => e
    Rails.logger.error("[channel error] #{e.message}")
    Rails.logger.error(e.backtrace[0..4].join("\n"))
    ActionCable.server.broadcast(
      client_key,
      {
        type: "error",
        message: "#{e.message} #{e.backtrace[0..4].join("\n  ")}",
      },
    )
  end

  private

  def send_operations_since(version)
    operations =
      @document.operations.where("version >= ?", version).order("version")
    Rails.logger.info(
      "send_operations_since(#{version}) => #{operations.length} ops",
    )
    # ActionCable.server.broadcast(
    #   client_key,
    #   { type: "op", message: operations.map(&:as_json) },
    # )
  end

  def broadcast_key
    "document:#{@document.id}:operations"
  end

  def client_key
    "document:#{@document.id}:operations_#{@client_id}"
  end
end

File: app/javascript/src/channels/DocumentChannel.ts
import { DocumentEntity } from './../store/document'
import { Action } from '@reduxjs/toolkit'
import consumer from './consumer'

/* eslint-disable no-this */
function channelCallbacks(
  channelClass: DocumentChannel,
  onConnected: () => void,
  onDisconnected: () => void
) {
  return {
    connected: () => {
      channelClass.log('[channelCallbacks] connected')
      onConnected()
    },
    disconnected: () => {
      channelClass.log('[channelCallbacks] disconnected')
      onDisconnected()
    },
    rejected: () => {
      channelClass.log('[channelCallbacks] rejected')
    },
    received: (
      data: AcknowledgementMessage | OperationMessage | ErrorMessage
    ) => {
      switch (data.type) {
        case 'op':
          return channelClass.onOperation(data, data.client_id)
        case 'ack':
          return channelClass.onAcknowledgement(data)
        case 'error':
          return channelClass._error(data.message)
        default:
          this.warn(`Can't process message with unrecognized type`, data)
      }
    },
  }
}

type OperationAPI = {
  id: number
  document_id: number
  operation_type: Action['type']
  payload: Operation
  version: number
  created_at: string
  updated_at: string
}

type AcknowledgementMessage = {
  type: 'ack'
  message: OperationAPI
}

type OperationMessage = {
  type: 'op'
  message: OperationAPI[]
  client_id: string
}

type ErrorMessage = {
  type: 'error'
  message: string
}

type Remoteish = {
  _remote: boolean
}

export type InsertOperation = {
  type: 'document/insert'
  payload: DocumentEntity & Remoteish
}

export type RemoveOperation = {
  type: 'document/remove'
  payload: Pick<DocumentEntity, 'id'> & Remoteish
}

// same as Redux PayloadAction<any> for all the actions we care about
export type Operation = InsertOperation | RemoveOperation

export type VersionedOperation = Operation & {
  version: number
}

export default class DocumentChannel {
  channel = null

  clientId: string
  documentId: number
  receivedCallback: (operations: Operation[]) => void
  acknowledgedCallback: (operation: Operation) => void
  errorCallback: (message: string) => void

  constructor({ clientId, documentId, onReceived, onAcknowledged, onError }) {
    this.clientId = clientId
    this.documentId = documentId
    this.receivedCallback = onReceived
    this.acknowledgedCallback = onAcknowledged
    this.errorCallback = onError

    this.log('channel ready', this.clientId)
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  log(...args: any[]) {
    // eslint-disable-next-line no-console
    console.log('%c[DocumentChannel]', 'color: #cae', ...args)
  }

  connect(onConnected: () => void, onDisconnected: () => void) {
    this.log('connecting...')
    this.channel = consumer.subscriptions.create(
      {
        channel: 'DocumentChannel',
        client_id: this.clientId,
        document_id: this.documentId,
      },
      channelCallbacks(this, onConnected, onDisconnected)
    )
  }

  submit(action: VersionedOperation) {
    this.channel.perform('submit', action)
  }

  requestOperationsSince(version: number) {
    this.channel.perform('request_operations_since', { version })
  }

  _error(data) {
    this.errorCallback && this.errorCallback(data)
  }

  onOperation(data: OperationMessage, clientId: string) {
    // ignore our own operations
    if (clientId === this.clientId) return
    this.receivedCallback &&
      this.receivedCallback(data.message.map(this.toVersionedOperation))
  }

  toVersionedOperation(op: OperationAPI): VersionedOperation {
    return {
      type: op.operation_type,
      payload: op.payload,
      version: op.version,
    }
  }

  onAcknowledgement(data: AcknowledgementMessage) {
    this.log('DocumentChannel ack', data)
    this.acknowledgedCallback &&
      this.acknowledgedCallback(this.toVersionedOperation(data.message))
  }
}
